qiskit/extensions/quantum_initializer/initializer.py

```diff
@@ -270,39 +267,76 @@ def validate_parameter(self, parameter):
                                "{1}".format(type(parameter), self.name))
 
 
-class InitializeGate(Gate):
-    """Complex amplitude initialization
+class Initialize(Instruction):
+    """Complex amplitude initialization.
 
     Class that implements the (complex amplitude) initialization of some
     flexible collection of qubit registers (assuming the qubits are in the
     zero state).
-    Note that InitializeGate is an Gate and not an Instruction since it does not contains
-    a reset instruction, contained in Initialize
+    Note that Initialize is an Instruction and not a Gate since it contains a reset instruction,
+    which is not unitary.
     """
 
-    def __init__(self, params, label=None):
-        """Create new initialize gate.
+    def __init__(self, params):
+        """Create new initialize composite.
 
         params (list): vector of complex amplitudes to initialize to
-        label: An optional label for the gate.
         """
+        try:
+            num_qubits = math.log2(len(params))
+        except:
+            raise QiskitError("Given params are not a statevector")
 
-        if not isinstance(params, list):
-            raise QiskitError("Given param is not a statevector")
-
-        num_qubits = math.log2(len(params))
-
+        # Check if param is a power of 2
         if num_qubits == 0 or not num_qubits.is_integer():
             raise QiskitError("Desired statevector length not a positive power of 2.")
 
-        super().__init__('init', int(num_qubits), params, label=label)
+        # Check if probabilities (amplitudes squared) sum to 1
+        if not math.isclose(sum(np.absolute(params) ** 2), 1.0,
+                            abs_tol=_EPS):
+            raise QiskitError("Sum of amplitudes-squared does not equal one.")
+
+        num_qubits = int(num_qubits)
+
+        super().__init__("initialize", num_qubits, 0, params)
 
     def _define(self):
-        desired_vector = self.params
+        """Calculate a subcircuit that implements this initialization
+
+        Implements a recursive initialization algorithm, including optimizations,
+        from "Synthesis of Quantum Logic Circuits" Shende, Bullock, Markov
+        https://arxiv.org/abs/quant-ph/0406176v5
+
+        Additionally implements some extra optimizations: remove zero rotations and
+        double cnots.
+        """
```

Comment:
I think this comment can be removed, it applies to `InitializeGate` since the decomposition is computed there.

Reply:
The comment is now removed